<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.
Use the version as per your python version.


In [132]:
!pip install yfinance
!pip install beautifulsoup4
!pip install nbformat
!pip install lxml


In [133]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [134]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [135]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard. 
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [136]:
tsla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [137]:
tsla_data = tsla.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [138]:
tsla_data.reset_index(inplace=True)
print(tsla_data.head())


                       Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0  


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [139]:
url =" https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm "
response = requests.get(url)
html_data = requests.get(url).content  # Use .content here

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [140]:
soup = BeautifulSoup(html_data, 'html.parser')


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/html/parser.py:171: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  k = self.parse_starttag(i)


Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [3]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

# Step 1: Create an empty DataFrame
tesla_revenue = pd.DataFrame(columns=["Date", "Revenue"])

# Step 2: Download the webpage
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/tesla_data_webpage.html"
html_data = requests.get(url).content

# Step 3: Parse the HTML
soup = BeautifulSoup(html_data, 'html.parser')

# Debug: Print out the first 500 characters of the HTML to understand its structure
print(soup.prettify()[:500])  # To inspect the content

# Step 4: Find all table bodies to see how many there are
table_bodies = soup.find_all("tbody")

# Debug: Print out how many table bodies we found
print(f"Found {len(table_bodies)} table bodies.")

# If there are enough tables, proceed to find the correct one
if len(table_bodies) > 1:
    table = table_bodies[1]  # Find the second table body
    # Step 5: Iterate through the rows in the table body
    for row in table.find_all("tr"):
        cols = row.find_all("td")
        if len(cols) >= 2:
            # Extract date and revenue
            date = cols[0].text.strip()
            revenue = cols[1].text.strip()

            # Step 6: Append the data to the DataFrame
            tesla_revenue = tesla_revenue.append({"Date": date, "Revenue": revenue}, ignore_index=True)

    # Display the first 5 rows of the DataFrame
    print(tesla_revenue.head())
else:
    print("No sufficient tables found. Please check the HTML structure.")


<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<error>
 <code>
  NoSuchKey
 </code>
 <message>
  The specified key does not exist.
 </message>
 <resource>
  /cf-courses-data/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/tesla_data_webpage.html
 </resource>
 <requestid>
  7a66809c-776d-4a44-b77f-104777a8292d
 </requestid>
 <httpstatuscode>
  404
 </httpstatuscode>
</error>

Found 0 table bodies.
No sufficient tables found. Please check the HTML structure.


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/html/parser.py:171: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  k = self.parse_starttag(i)


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [4]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

<>:1: SyntaxWarning: invalid escape sequence '\$'
<>:1: SyntaxWarning: invalid escape sequence '\$'
/var/folders/mb/6p3wtb813bg2nkf0321y44k80000gn/T/ipykernel_96057/349343550.py:1: SyntaxWarning: invalid escape sequence '\$'
  tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")


Execute the following lines to remove an null or empty strings in the Revenue column.


In [5]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [6]:
print(tesla_revenue.tail())

Empty DataFrame
Columns: [Date, Revenue]
Index: []


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [8]:
import yfinance as yf
import pandas as pd
from bs4 import BeautifulSoup
gme = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [ ]:
gme_data = gme.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [14]:
gme_data = pd.DataFrame()
gme_data.reset_index(inplace=True)
print(gme_data.head())


Empty DataFrame
Columns: [index]
Index: []


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [15]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
html_dat2  = requests.get(url).text

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [20]:
soup = BeautifulSoup(html_dat2,'html.parser')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [35]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

# Step 1: Create an empty DataFrame
gme_revenue = pd.DataFrame(columns=["Date", "Revenue"])

# Step 2: Download the webpage
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/tesla_data_webpage.html"
html_data2 = requests.get(url).content

# Step 3: Parse the HTML
soup = BeautifulSoup(html_data, 'html.parser')

# Debug: Print out the first 500 characters of the HTML to understand its structure
print(soup.prettify()[:500])  # To inspect the content

# Step 4: Find all table bodies to see how many there are
table_bodies = soup.find_all("tbody")

# Debug: Print out how many table bodies we found
print(f"Found {len(table_bodies)} table bodies.")

# If there are enough tables, proceed to find the correct one
if len(table_bodies) > 1:
    table = table_bodies[1]  # Find the second table body
    # Step 5: Iterate through the rows in the table body
    for row in table.find_all("tr"):
        cols = row.find_all("td")
        if len(cols) >= 2:
            # Extract date and revenue
            date = cols[0].text.strip()
            revenue = cols[1].text.strip()

            # Step 6: Append the data to the DataFrame
            gme_revenue = gme_revenue.append({"Date": date, "Revenue": revenue}, ignore_index=True)
            tesla_revenue.dropna(inplace=True)



    # Display the first 5 rows of the DataFrame
    print(gme_revenue.head())
else:
    print("No sufficient tables found. Please check the HTML structure.")
gme_revenue = tesla_revenue[gme_revenue['Revenue'] != ""]    


<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<error>
 <code>
  NoSuchKey
 </code>
 <message>
  The specified key does not exist.
 </message>
 <resource>
  /cf-courses-data/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/tesla_data_webpage.html
 </resource>
 <requestid>
  457a2dea-97e4-4edb-bc39-07ab33378c2f
 </requestid>
 <httpstatuscode>
  404
 </httpstatuscode>
</error>

Found 0 table bodies.
No sufficient tables found. Please check the HTML structure.


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/html/parser.py:171: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  k = self.parse_starttag(i)


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [36]:
print(gme_revenue.tail())

Empty DataFrame
Columns: [Date, Revenue]
Index: []


## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [39]:
import yfinance as yf
import plotly.graph_objects as go

# Download Tesla stock data from Yahoo Finance
tesla_data = yf.download('TSLA', start='2010-01-01', end='2021-06-30')

# Create the figure for the graph
fig = go.Figure()

# Add a trace for the Tesla stock closing price
fig.add_trace(go.Scatter(x=tesla_data.index, y=tesla_data['Close'], mode='lines', name='Tesla Close Price'))

# Add a title to the graph
fig.update_layout(title="Tesla Stock Data (Up to June 2021)",
                  xaxis_title="Date",
                  yaxis_title="Stock Price (USD)")

# Show the graph
fig.show()


[*********************100%***********************]  1 of 1 completed


## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [42]:
import yfinance as yf
import plotly.graph_objects as go

# Download GameStop stock data from Yahoo Finance
gme_data = yf.download('GME', start='2010-01-01', end='2021-06-30')

# Define the make_graph function
def make_graph(stock_data, revenue_data, stock_name):
    fig = go.Figure()

    # Add a trace for the stock closing price
    fig.add_trace(go.Scatter(x=stock_data.index, y=stock_data['Close'], mode='lines', name=f'{stock_name} Close Price'))

    # You can add other traces for the revenue data if needed
    # For example, if gme_revenue has columns Date and Revenue, you can add that as:
    # fig.add_trace(go.Bar(x=revenue_data['Date'], y=revenue_data['Revenue'], name=f'{stock_name} Revenue'))

    # Add a title to the graph
    fig.update_layout(title=f"{stock_name} Stock Data (Up to June 2021)",
                      xaxis_title="Date",
                      yaxis_title="Stock Price (USD)")

    # Show the graph
    fig.show()

# Call the function with gme_data, gme_revenue, and the stock name 'GameStop'
make_graph(gme_data, gme_revenue, 'GameStop')


[*********************100%***********************]  1 of 1 completed


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
